storage/internal/benchmarks/client_pool.go

```diff
@@ -31,130 +30,125 @@ import (
 	"google.golang.org/grpc"
 )
 
-// clientPool functions much like a sync Pool (https://pkg.go.dev/sync#Pool),
-// except it does not automatically remove items stored in the clientPool.
-// Re-using the clients rather than creating a new one each time reduces overhead
-// (such as re-creating the underlying HTTP client and opening credential files),
-// and is the intended way to use Storage clients.
-//
-// There is no limit to how many clients will be created, but it should be around
-// the order of 5 * min(workers, max_samples).
+// clientPool pools a number of Storage clients for use without blocking, ie.
+// a client that is received through Get may still be in use by one or more other
+// calls.
 type clientPool struct {
-	New     func() *storage.Client
-	clients []*storage.Client
+	clients     []*storage.Client
+	clientQueue chan *storage.Client
 }
 
-func (p *clientPool) Get() *storage.Client {
-	// Create the slice if not already created
-	if p.clients == nil {
-		p.clients = make([]*storage.Client, 0)
+// newClientPool initializes the pool with numClients clients initialized using
+// the initializeClient func.
+// Returns the client pool and a cleanup func to be called to close the pool.
+func newClientPool(initializeClient func() (*storage.Client, error), numClients int) (*clientPool, func()) {
+	p := &clientPool{
+		clients:     make([]*storage.Client, numClients),
+		clientQueue: make(chan *storage.Client, numClients),
 	}
 
-	// If there is an unused client, return it
-	if len(p.clients) > 0 {
-		c := p.clients[0]
-		p.clients = p.clients[1:]
-		return c
-	}
+	for i := 0; i < numClients; i++ {
+		var err error
+		p.clients[i], err = initializeClient()
+		if err != nil {
+			log.Fatalf("initializeClient: %v", err)
+		}
 
-	// Otherwise, create a new client and return it
-	return p.New()
-}
+		// Fill the queue with clients as they are created
+		p.clientQueue <- p.clients[i]
+	}
 
-func (p *clientPool) Put(c *storage.Client) {
-	p.clients = append(p.clients, c)
+	return p, func() {
+		for _, c := range p.clients {
+			c.Close()
+		}
+	}
 }
 
-// we can share clients as long as the app buffer sizes are constant
-var httpClients, gRPCClients *clientPool
+// Get rotates through clients. This means the work may not be evenly distributed,
+// particularly if using varying object sizes.
+func (p *clientPool) Get() *storage.Client {
+	client := <-p.clientQueue
 
-var nonBenchmarkingClients = clientPool{
-	New: func() *storage.Client {
-		// For debuggability's sake, these are HTTP
-		clientMu.Lock()
-		client, err := storage.NewClient(context.Background())
-		clientMu.Unlock()
-		if err != nil {
-			log.Fatalf("storage.NewClient: %v", err)
-		}
+	// return client to queue so that it will be used again without blocking
+	p.clientQueue <- client
```

Comment:
Might want to note in a comment that this code is not totally deterministic if called from multiple threads at the same time -- you could have clients swap places in the queue potentially. But, I don't think it's necessary to add a mutex since this is presumably acceptable.

Reply:
I added a comment. I think it is acceptable. If we are concerned with unbalanced workloads, we may implement a priority queue as mentioned in the other comment.

Though, adding a mutex here is probably low-cost?

Reply:
Agreed, I think it's fine, just good to have a comment for anyone looking at this code in the future.